### ISSUE TITLE
[BUG] Installing addon from GitHub branch can't find default blueprint

### ISSUE THREAD

**Issue by srvance** [ember-cli/ember-cli#6833]
Run

```command
ember install "https://github.com/kaliber5/ember-bootstrap#bs4-blueprint"
```

You will get the following warning line in the output.

```
WARNING: Could not figure out blueprint name from: "https://github.com/kaliber5/ember-bootstrap#bs4-blueprint". Please install the addon blueprint via "ember generate <addon-name>" if necessary.
```

This comes from [here](https://github.com/ember-cli/ember-cli/blob/master/lib/tasks/addon-install.js#L66-L67). It ultimately traces down to the code in `npm-package-arg` for test cases like [this one](https://github.com/npm/npm-package-arg/blob/master/test/github.js#L54-L60) where the resulting name from parsing remote references is `null`.

It seems like an incomplete implementation on the part of the dependency, but I don't feel like I have a complete enough view of the usage context to be sure.

---

Output from `ember version --verbose && npm --version`:
```
[Replace this line with the output.]
```

**Comment by Turbo87** [comment]
> It seems like an incomplete implementation on the part of the dependency, but I don't feel like I have a complete enough view of the usage context to be sure.

the problem is that there is no way to figure out the name of the npm package just by looking at the git path. e.g. the `name` in `package.json` could be different from the repo name.

**Comment by stefanpenner** [comment]
@Turbo87 ya, it seems in-error that we do not treat the name in package.json as the source of truth here. An exception might be scoped packages, where the scope is from disk, but the name is still from package.json

**Comment by Turbo87** [comment]
That's not the point. We have no way to actually find the right package.json without guessing.

**Comment by srvance** [comment]
How likely is it that someone who knows enough to use a URL for their dependency wouldn't have a `package.json` at the root? If someone was just doing it because they were told to, the same would be true of the person who told them. Is there a scenario I don't understand there?

It seems more correct than current to try to use the repo name and more correct than that to use the name from the root `package.json`. But `npm-package-arg` just seems to parse the reference anyway.

**Comment by stefanpenner** [comment]
> That's not the point. We have no way to actually find the right package.json without guessing.

tell me more

**Comment by Turbo87** [comment]
> > That's not the point. We have no way to actually find the right package.json without guessing.
>
> tell me more

if you use 

```
ember install "https://github.com/kaliber5/ember-bootstrap#bs4-blueprint"
```

we run

```
npm install --save-dev "https://github.com/kaliber5/ember-bootstrap#bs4-blueprint"
```

which works fine and npm will install the package. But where did npm install the package? for stuff like `npm install ember-cli` it is obviously `node_modules/ember-cli` since the package name in the npm registry matches the folder name, but for URLs this is not obvious.

we could **guess** that in the above case it will be `node_modules/ember-bootstrap`, but we can not be certain because npm does not return the paths of the installed packages.

**Comment by stefanpenner** [comment]
@Turbo87 ah, that makes sense. Thanks.

**Comment by les2** [comment]
if run interactively, ember-cli could prompt the user to enter the add-on name. the default value could be a guessed value.

when guessing, ember-cli could see if it's possible to:
1. resolve package.json from the guessed name
2. verify that the package is indeed an ember addon (looking for ember addon entry in package json)

### alternative

add cli option to specify package name to use when installing from git url.

### other thoughts
the current warning could coach the user a little more. perhaps it could print a look to a section of the guide that covers this caveat.

as an experienced cli user i would understand what the warning means. i think a lot of developers don't necessarily know that addons typically install blueprints once during install phase and why this is important

**Comment by les2** [comment]
i don't think this is a bug but rather an enhancement request.

**Comment by Turbo87** [comment]
I guess we could try to do the right thing by always running `npm install`, checking if a node modules exists with a name matching the repo name and otherwise print a helpful warning that the blueprint has to be run manually via `ember generate <whatever-the-package-name-is>`

**Comment by srvance** [comment]
> But where did npm install the package?

What are the options and influencers on this?

**Comment by Turbo87** [comment]
@srvance it will be installed into `node_modules` with the folder name matching the `name` property in the `package.json` file.

but to get to the `package.json` file we need the folder name, and to get to the folder name we need the `package.json` file, and to get to the `package.json` file we need the folder name, and to get to the folder name we need the `package.json` file, and to get to the `package.json` file we need the folder name, and to get to the folder name we need the `package.json` file, and to get to the `package.json` file we need the folder name, and to get to the folder name we need the `package.json` file, and to get to the `package.json` file we need the folder name, and to get to the folder name we need the `package.json` file, ...

you see where this is going

**Comment by srvance** [comment]
LOL. Got it. And I can't find an npm command that gives that local information, either. Although we could probably have a high hit rate if we try the name parsed out of the URL to reduce the warning occurrence.

**Comment by Turbo87** [comment]
> Although we could probably have a high hit rate if we try the name parsed out of the URL to reduce the warning occurrence.

yeah, that would probably be a good solution for the 90% case if we explicitly warn in the other 10% of cases.

**Comment by srvance** [comment]
And there's already a good warning.

**Comment by hjdivad** [comment]
after `npm install 'https://github.com/kaliber5/ember-bootstrap#bs4-blueprint'`

```
[N] ⋊> ~ ls -d ~/.npm/_git-remotes/git-https*
/Users/dhamilto/.npm/_git-remotes/git-https-github-com-kaliber5-ember-bootstrap-git-bs4-blueprint-aff490f9

[N] ⋊> ~ cd /Users/dhamilto/.npm/_git-remotes/git-https-github-com-kaliber5-ember-bootstrap-git-bs4-blueprint-aff490f9; and git show bs4-blueprint:package.json | jq .name
"ember-bootstrap"
```

Unsure across what versions `$HOME/.npm/_git-remotes` can be relied upon, but it looks like there is a reliable, if tedious, way to get the package name

**Comment by Turbo87** [comment]
That looks like very private "API" to me and I'd rather not use that. And for yarn this would need yet another approach, so not really a great universal solution.

**Comment by les2** [comment]
is looking for a `package.json` in the git repo itself reasonable? it would imply cloning the repo (again) to a temp folder and looking for the file or maybe requesting just that file from the remote.

so that would give the package.json, which would give the package name, which would allow running the blueprint.

**Comment by Turbo87** [comment]
@les2 that would be a solution, but that's quite a lot of overhead and complexity that I'd rather avoid. we've discussed this in the weekly CLI call yesterday and came to the conclusion that we should "guess" the module name from the URL if possible and if such a module does not exist we will show a warning.

this seems to be actionable now and reasonably well described so anyone that's interested feel free to open a PR for this :)

**Comment by srvance** [comment]
Any opinion whether  logic should be in ember-cli or npm-package-arg? It seems it should belong in the latter, but for the guessing part.

**Comment by Turbo87** [comment]
@srvance this should definitely belong in Ember CLI itself, since we're only *guessing* the project name.

**Comment by stefanpenner** [comment]
What if in the cases where we cannot infer we asked the user to tell us via command line arg/flag?

Downsides are that it is extra work, and people may typo it...

---

Alternatively we could download the repo ourselves then check the name and move it.

This may also be important for when installing from a gist, or directory on disk, or tarball

**Comment by srvance** [comment]
@stefanpenner 

> What if in the cases where we cannot infer we asked the user to tell us via command line arg/flag?

Personally, I prefer that interactive usage be opt-in so that the default mode is scriptability. If we decide that interaction can be the default, we need to have a way to turn off all interactive mode prompts for scripting. We're having a similar discussion around an interactive mode for the ember-bootstrap blueprint to get around emberjs/rfcs#212.

**Comment by michielboekhoff** [comment]
I've picked up this issue - one thing I am decidedly stuck on, is that the expected behaviour for `ember install "ember-cli/ember-cli-qunit"` according to the tests is **not** to install the addon blueprint. However, is this still desirable or shall I change/remove this test?

**Comment by mlwilkerson** [comment]
Hi @michielboekhoff. I'm facing this problem now as well: creating an addon that has a scoped package name (like: @account/foo-bar), and ember install can't figure out the default blueprint to run. Are you still working on this? Maybe my situation could help to shed some light on it.
